### DISCUSSION TITLE
Build a template to add external sensor on CC2650 Launchpad in cc26x0-web-demo example

### DISCUSSION THREAD

**Discussion by atulkprajapati** [contiki-ng/contiki-ng#1601] [General]
I'm planning to build a template with relevant functions which can help read any external sensor that we wish to add on a CC2650 Launchpad, since I'm working with cc26x0-web-demo example for a WSN, I also plan to add a coap resource to it, to read the sensor readings using coap commands. For this I'm unable to understand where do I start it, because for example we have been provided with dht11-sensor.c([here](https://github.com/contiki-ng/contiki-ng/tree/develop/arch/dev/sensor/dht11)), but I'm unable to figure out how can I actually use it for my launchpad. So I need some guidance, as to how can I make a general template, where-in, we just need to place the relevant code for that particular sensor, and voila, it starts to work as intended! So looking forward for your valuable guidance.

**Comment by Yagoor** [comment]
Hi, the first step is to try the [example](https://github.com/contiki-ng/contiki-ng/tree/develop/examples/dev/dht11) on your board. The DHT11 is a bit bang peripheral so timing is critical here. Once you have tested the example and you got results. You can try to integrate the sensor into your application.

**Reply by atulkprajapati** [reply]
What all things do I need to import, to test it on CC2650 Launchpad, I saw an example on youtube, wherein it was done on MSP430, it was simply just to include a <msp430.h> header file, which got auto included when choosing the board in CCS. Nothing as such happens when trying with CC2650.

**Reply by Yagoor** [reply]
You can just built the example above targeting the CC2650 Launchpad. I'm not sure on which TARGET and BOARD you need to use for that.

**Reply by atulkprajapati** [reply]
I know those are:
TARGET = cc26x0-cc13x0
BOARD = launchpad/cc2650

Using the make command I got the bin file, which I flashed on the device,  how to proceed next, to check for the values ?

**Reply by Yagoor** [reply]
I'm not familiar with this platform but it must have some serial interface, either USB or UART to get logs out of it.
Once you know if it is USB or UART that this platform supports you can use some terminal program to connect to it like putty on windows or pyserial which should work on all os variants.

**Reply by atulkprajapati** [reply]
It does have a UART module ([here](https://www.ti.com/lit/ug/swcu117i/swcu117i.pdf?ts=1624343730326) page 1452), that probably requires a [Converter](https://www.segger.com/products/connectivity/emusb-host/add-ons/cp210x/) which i don't have for now. Although the board gets connected to the PC using a USB(if that is what you meant by USB)

**Reply by Yagoor** [reply]
I think this platform has the USB CDC, so you can probably also get the logs from the USB else you'll need the UART.

**Reply by atulkprajapati** [reply]
I'm able to get the logs on the serial terminal now! It just comes and stops here.
![image](https://user-images.githubusercontent.com/22424087/123224266-4bdd5200-d4ef-11eb-9b2c-2e56669ba377.png)

**Reply by Yagoor** [reply]
That is good. Now you need to use the example I mentioned above and see if you get the sensor readings.

**Reply by atulkprajapati** [reply]
I used this command to generate the .elf file
make TARGET=cc26x0-cc13x0 BOARD=launchpad/cc2650 dht11.upload PORT=/dev/ttyACM2
Then read using: 
sudo make login PORT=/dev/ttyACM2
![image](https://user-images.githubusercontent.com/22424087/123384158-b1921280-d5b1-11eb-80b1-8abc66b9780f.png)
I have been getting these timed out, the only thing I'm confused is about the PIN & PORT number configuration in line No. 49 & 50 of [dht11.c](https://github.com/contiki-ng/contiki-ng/blob/develop/examples/dev/dht11/dht11.c). 

Can you please suggest me the PIN & PORT according to this architecture: [CC2650 Launchpad](https://www.ti.com/lit/ml/swru451/swru451.pdf)

![image](https://user-images.githubusercontent.com/22424087/123386926-cde37e80-d5b4-11eb-9f41-00c9119611ac.png)

Thank you for always being prompt with your help.

**Reply by Yagoor** [reply]
I don't think this board has port only pins. So you need to change the pin to the pin in which you plugged the sensor on.

**Reply by atulkprajapati** [reply]
So just changing this to: 
#define DHT11_GPIO_PIN  (12)
To
#define DHT11_GPIO_PIN    IOID_22
![image](https://user-images.githubusercontent.com/22424087/123515588-a1fbf200-d6b5-11eb-8e1c-7be8d50126b4.png)

Also, remove that PORT parameter? Will do ?

**Reply by Yagoor** [reply]
I'm not sure on the pin out of your board. You can try the IOID_22

**Reply by atulkprajapati** [reply]
The C file has #include "dht11-sensor.h"
Is it [this](https://github.com/contiki-ng/contiki-ng/blob/develop/arch/dev/sensor/dht11/dht11-sensor.h) one 
1) So, that I can comment out the Port declaration ?

Also this is a [video](https://www.youtube.com/watch?v=Hid_jB_Dy-A&t=460s) for interfacing DHT11 in an MSP430 Microcontroller
2) In this setup there is a 10k ohm resistor, so it that also a part of testbed for the code that you've written?
![image](https://user-images.githubusercontent.com/22424087/123589275-a126a580-d806-11eb-8289-2f1f2748ec59.png)

3) Also, can you please specify which platform did you write this code for ?

**Reply by Yagoor** [reply]
1. You don't need to comment out the port declaration. IF your board does not have port, contiki will ignore it in the gpio-hal.
2. You need to find which IOID matches the P2.4.
3. The code was written for contiki-ng, so it should work with any port contiki-ng supports. The example code was tested on a nrf52840/dk board.

**Reply by atulkprajapati** [reply]
1. It asks for a parameter, else it throws error 
![image](https://user-images.githubusercontent.com/22424087/123804459-dddbc500-d90a-11eb-9b22-21cf5a7748a5.png)
2. It says, the MSP430 Pin configuration shows its Digital Read()/Write() & AnalogWrite()
The similar PIN in CC2650 is the DIO23

**Reply by Yagoor** [reply]
1. Instead of leaving it empty, just use 0. :)

**Reply by atulkprajapati** [reply]
With that, there's no response.
Unlike last time, at least it showed; Timed out!

This is something strange...what does this indicate. Although this happened only once.
![image](https://user-images.githubusercontent.com/22424087/124256585-19fc6900-db49-11eb-8360-3569dd473b25.png)



Then again, there is no response!
![image](https://user-images.githubusercontent.com/22424087/124257501-0d2c4500-db4a-11eb-91d2-a027719f4ddf.png)
